### PR TITLE
Add more extractors for commons and improve links between DBpedia and DBpedia Commons

### PR DESCRIPTION
**Pull request by MrSerth**
This PR includes the following improvements:

- Query the commons.wikimedia.org API instead of using a local image dump
- Extract as many media items (images, audio and video files) as possible and add links to the corresponding resources in commons.dbpedia.org (`MediaExtractor`) for regular DBpedia language versions (such as `en`, `de` or `fr`) and store them in a new `dbpedia-owl:mediaItem` property.
- Add default files for Commons live extraction (`live_commons.default.xml` and `live_commons.default.ini`) 
- Include the namespaces in `ExtractorUtils.commonsNamespacesContainingMetadata` for DBpedia Commons.
- `owl:sameAs` between DBpedia and DBpedia Commons (`CommonsResourceExtractor`) and backwards (`DBpediaResourceExtractor`)
- Ensure triples including references to commons.wikimedia.org don't link to the wrong commons.wikipedia.org domain
- Exclude non-free images (as those are not available on commons.wikimedia.org)

### CODE REVIEW

File: core/src/main/resources/addonlangs.json [hunk 1]
```diff
@@ -23,16 +23,22 @@
...
-    "name": "Core Directory",
-    "isoCode": "en",
-    "iso639_3": "eng"
-  },
   "commons": {
     "wikiCode": "commons",
     "name": "Commons",
     "isoCode": "en",
+    "iso639_3": "eng",
+    "dbpediaDomain":"commons.dbpedia.org",
```
Comment (MrSerth) [inline core/src/main/resources/addonlangs.json]: We mainly needed to include the `baseUri` and `apiUri` as they are different from the usual wikipedia URL (and should be commons.wikimedia.org). Otherwise, they would point to commons.wikipedia.org and so would the triples.

Comment (MrSerth) [inline core/src/main/resources/addonlangs.json]: We identified [this commit](https://github.com/dbpedia/extraction-framework/commit/e0e82f90acf147687412ef79a53f6b5bc4859b02#diff-25daade9308e37621b81033b0a9f4860L93) as the reason for the issue we ran into as the special commons handling was removed.

File: core/src/main/resources/addonlangs.json [hunk 2]
```diff
@@ -23,16 +23,22 @@
     "baseUri":"http://www.wikidata.org",
     "apiUri":"https://www.wikidata.org/w/api.php"
   },
-  "core": {
```
Comment (jimkont) [inline core/src/main/resources/addonlangs.json]: btw, why did you delete "core"? I think @chile12 added this for dataID

Comment (jimkont) [inline core/src/main/resources/addonlangs.json]: ignore :)

File: core/src/main/scala/org/dbpedia/extraction/mappings/CommonsResourceExtractor.scala [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ */
+class CommonsResourceExtractor (
+  context : {
+    def ontology : Ontology
+    def language : Language
+  }
+) extends PageNodeExtractor{
+
+  private val propertyUri = context.ontology.properties("owl:sameAs");
+  private val commonsBaseUri = "http://commons.dbpedia.org/resource/%s"
```
Comment (jimkont) [inline core/src/main/scala/org/dbpedia/extraction/mappings/CommonsResourceExtractor.scala]: I would use `Language.Commons.dbpediaUri` here

File: core/src/main/scala/org/dbpedia/extraction/mappings/CommonsResourceExtractor.scala [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import org.dbpedia.extraction.destinations.{DBpediaDatasets, Quad}
+import org.dbpedia.extraction.ontology.Ontology
+import org.dbpedia.extraction.util.Language
+import org.dbpedia.extraction.wikiparser.{PageNode, TextNode}
+
+import scala.collection.mutable.ArrayBuffer
+
+/**
+ * Links non-commons DBpedia resources to their DBpedia Commons counterpart using owl:sameAs.
+ * This requires the the Wikipedia page to contain a {{Commons}} template.
```
Comment (jimkont) [inline core/src/main/scala/org/dbpedia/extraction/mappings/CommonsResourceExtractor.scala]: can you also extend this a bit and show an example input template with values and the generated triples?

File: core/src/main/scala/org/dbpedia/extraction/mappings/CommonsResourceExtractor.scala [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    for { template <- InfoboxExtractor.collectTemplates(node)
+      if template.title.decoded == "Commons"
+    }
+    {
+      if (template.children.isEmpty){
+        return Seq(new Quad(context.language, DBpediaDatasets.PageLinks, subjectUri, propertyUri,
+          String.format(commonsBaseUri, node.title.encoded.asInstanceOf[String]), null, null))
+      } else{
+        var commonsPageTitle = template.children.head.children.head.asInstanceOf[TextNode].text
```
Comment (jimkont) [inline core/src/main/scala/org/dbpedia/extraction/mappings/CommonsResourceExtractor.scala]: I think `WikiTitle.parse` is safer to use, I'd also use a Wikititle object with an altered language in the former if satement

File: core/src/main/scala/org/dbpedia/extraction/mappings/DBpediaResourceExtractor.scala [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+import org.dbpedia.extraction.destinations.{DBpediaDatasets, Dataset, Quad}
+import org.dbpedia.extraction.ontology.{Ontology, OntologyProperty}
+import org.dbpedia.extraction.util.{Language, WikiUtil}
+import org.dbpedia.extraction.wikiparser.{Namespace, PageNode, PropertyNode, TextNode}
+import scala.language.reflectiveCalls
+
+/**
+ * Links DBpedia Commons resources to their counterparts in other DBpedia languages (only en, de and fr) using owl:sameAs.
+ * This requires the the Wikimedia page to contain a {{VN}} template.
```
Comment (jimkont) [inline core/src/main/scala/org/dbpedia/extraction/mappings/DBpediaResourceExtractor.scala]: again, here, can you add a complete example?

File: core/src/main/scala/org/dbpedia/extraction/mappings/CommonsResourceExtractor.scala (hunk 5 above)
Comment (jimkont) [inline core/src/main/scala/org/dbpedia/extraction/mappings/CommonsResourceExtractor.scala]: if you use `WikiTitle` you'll probably won't need the `commonsBaseUri` variable

File: core/src/main/scala/org/dbpedia/extraction/mappings/DBpediaResourceExtractor.scala [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+      if (template.title.decoded == "VN")
+    } {
+      return template.children
+        .filter((node : PropertyNode) => Seq("de", "en", "fr").contains(node.key))
+        .map((node : PropertyNode) =>
+          new Quad(
+            context.language,
+            DBpediaDatasets.PageLinks,
+            subjectUri, propertyUri,
+            String.format(objectBaseUri, if (node.key == "en") "" else node.key + ".", WikiUtil.wikiEncode(node.children.head.asInstanceOf[TextNode].text.split(", ").head)), null, null
```
Comment (jimkont) [inline core/src/main/scala/org/dbpedia/extraction/mappings/DBpediaResourceExtractor.scala]: Should use WikiTitle here again, also, hiding the `en.` sub-domain is done automatically at later steps, no need to hardcode it here

File: core/src/main/scala/org/dbpedia/extraction/util/ExtractorUtils.scala [hunk 8]
```diff
@@ -114,6 +114,18 @@ object ExtractorUtils {
...
   /**
+    * Determine the file URL on DBpedia for a filename.
+    *
+    * @param filename the name of the file.
+    * @param language the wiki on which the file exists.
+    * @return the file URL
+    */
+  def getDbpediaFileURL(filename: String, language: Language): String = {
+    val fileNamespaceIdentifier = Namespace.File.name(language)
+    language.dbpediaUri + "/resource/" + fileNamespaceIdentifier + ":" + filename
```
Comment (jimkont) [inline core/src/main/scala/org/dbpedia/extraction/util/ExtractorUtils.scala]: should use `language.resourceUri` here.
about `filename` is this clean? otherwise you can use `WikiTitle.parse` again

File: live/src/main/java/org/dbpedia/extraction/live/queue/LiveQueue.java [hunk 9]
```diff
@@ -23,7 +23,7 @@
 
     private static final PriorityBlockingQueue<LiveQueueItem> queue = new PriorityBlockingQueue<LiveQueueItem>(1000);
     // this is a Map<long,int>, long is the pageID and int the number of same items in the queue (see add())
-    private static final Set<Long> uniqueSet = new HashSet(1000);
+    private static final Set<String> uniqueSet = new HashSet(1000);
```
Comment (jimkont) [inline live/src/main/java/org/dbpedia/extraction/live/queue/LiveQueue.java]: IIRC, Before RCStream not all items had a title, now not all items have a pageID so I am not sure if this change will work

File: live/src/main/scala/org/dbpedia/extraction/config/mappings/MediaExtractorConfig.scala [hunk 10]
```diff
@@ -0,0 +1,25 @@
+package org.dbpedia.extraction.config.mappings
```
Comment (jimkont) [inline live/src/main/scala/org/dbpedia/extraction/config/mappings/MediaExtractorConfig.scala]: Why is this not in `core` module?

File: live/src/main/scala/org/dbpedia/extraction/mappings/MediaExtractor.scala [hunk 11]
```diff
@@ -0,0 +1,160 @@
+package org.dbpedia.extraction.mappings
```
Comment (jimkont) [inline live/src/main/scala/org/dbpedia/extraction/mappings/MediaExtractor.scala]: same with config, why not move this to core, is it specific to live only?

**Comment by jimkont** [comment]
Thanks guys! looks quite good!
Made a first round of review & sorry for the delay, 
would be faster to merge if you kept the PRs smaller and more focused but this looks like a great contribution!

File: live/src/main/java/org/dbpedia/extraction/live/queue/LiveQueue.java (hunk 9 above)
Comment (MrSerth) [inline live/src/main/java/org/dbpedia/extraction/live/queue/LiveQueue.java]: We adjusted the `OAIFeederMappings` and the `UnmodifiedFeeder` to ensure every `LiveQueueItem` has a title. There isn't any use of the smaller `LiveQueueItem` constructor (without passing a title) any more for an object that is added to the `LiveQueue`.

**Comment by MrSerth** [comment]
Thanks for your comments and suggestions (and sorry for our delay)!
We tried to incorporate your ideas, please have another look.

P.S.: Our next PR will be much smaller with less changes...

File: core/src/main/scala/org/dbpedia/extraction/mappings/CommonsResourceExtractor.scala [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+    val quads = new ArrayBuffer[Quad]()
+
+    for { template <- InfoboxExtractor.collectTemplates(node)
+      if template.title.decoded == "Commons"
+    }
+    {
+      if (template.children.isEmpty){
+        val commonsResourceURL = WikiTitle.parse(node.title.encoded.asInstanceOf[String], commonsLanguage).resourceIri
+        return Seq(new Quad(context.language, DBpediaDatasets.PageLinks, subjectUri, propertyUri, commonsResourceURL, null, null))
```
Comment (jimkont) [inline core/src/main/scala/org/dbpedia/extraction/mappings/CommonsResourceExtractor.scala]: `DBpediaDatasets.PageLinks` is not the appropriate dataset for these triples, can you create a new one? Something like `CommonsLinks` or similar would do

File: core/src/main/scala/org/dbpedia/extraction/mappings/ContributorExtractor.scala [hunk 13]
```diff
@@ -30,7 +30,7 @@ class ContributorExtractor( context : {
 
     if(node.contributorID <= 0) return Seq.empty
 
-    val pageURL = "http://" + context.language.wikiCode + ".wikipedia.org/wiki/" + node.title.encoded;
+    val pageURL = context.language.baseUri + "/wiki/" + node.title.encoded;
```
Comment (jimkont) [inline core/src/main/scala/org/dbpedia/extraction/mappings/ContributorExtractor.scala]: Can you save the language in a class variable? We had some issues when we were accessing `context ` members inside the `extract()` function from Spark

File: core/src/main/scala/org/dbpedia/extraction/mappings/DBpediaResourceExtractor.scala [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    for {
+      template <- InfoboxExtractor.collectTemplates(page)
+      if template.title.decoded == "VN"
+    } {
+      return template.children
+        .filter((node : PropertyNode) => Seq("de", "en", "fr").contains(node.key))
+        .map((node : PropertyNode) =>
+          new Quad(
+            context.language,
+            DBpediaDatasets.PageLinks,
```
Comment (jimkont) [inline core/src/main/scala/org/dbpedia/extraction/mappings/DBpediaResourceExtractor.scala]: You can use the same dataset as above here

File: core/src/main/scala/org/dbpedia/extraction/mappings/MediaExtractor.scala [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+  {
+    if(node.title.namespace != Namespace.Main) return Seq.empty
+
+    var quads = new ArrayBuffer[Quad]()
+
+    val api = new WikiApi(null, language)
+    var firstImage = true
+    for ((mediaFileName, sourceNode) <- searchMedia(node.children, 0).flatten
+         if (!mediaFileName.toLowerCase.startsWith("replace_this_image"))
+           && (api.fileExistsOnWiki(mediaFileName, commonsLang)))
```
Comment (jimkont) [inline core/src/main/scala/org/dbpedia/extraction/mappings/MediaExtractor.scala]: This call here will probably create too many API calls on Wikimedia

File: live/src/main/java/org/dbpedia/extraction/live/storage/DBpediaSQLQueries.scala [hunk 16]
```diff
@@ -34,7 +34,7 @@ object DBpediaSQLQueries {
   }
 
   def getJSONCacheUnmodified: String = {
-    "SELECT pageID, updated FROM DBPEDIALIVE_CACHE WHERE updated <= now() - INTERVAL ? DAY ORDER BY updated ASC LIMIT ? "
+    "SELECT pageID, title, updated FROM DBPEDIALIVE_CACHE WHERE updated <= now() - INTERVAL ? DAY ORDER BY updated ASC LIMIT ? "
```
Comment (jimkont) [inline live/src/main/java/org/dbpedia/extraction/live/storage/DBpediaSQLQueries.scala]: The cache is initialized with pageID's only on the first run and some titles can be empty/null

**Comment by jimkont** [comment]
Hey again guys, I made another round of comments, very minor this time.

You should also add the new extractors in the [dump config file ](https://github.com/dbpedia/extraction-framework/blob/master/dump/extraction.default.properties#L61) (except the one that does the api calls)

**Comment by jimkont** [comment]
Thanks a lot guys! btw ping me offline to give you access to our slack channel if you want

File: live/src/main/java/org/dbpedia/extraction/live/storage/DBpediaSQLQueries.scala (hunk 16 above)
Comment (MrSerth) [inline live/src/main/java/org/dbpedia/extraction/live/storage/DBpediaSQLQueries.scala]: I opened a new PR #479 for this issue.
